data/tabular/train_test_split.py

```diff
@@ -0,0 +1,141 @@
+import tdc
+import pandas as pd
+import numpy as np
+from typing import List, Tuple, Dict, Union, Optional
+import sys
+from tqdm import tqdm
+from random import Random
+from collections import defaultdict
+from rdkit import Chem
+from rdkit.Chem.Scaffolds import MurckoScaffold
+from rdkit import RDLogger
+from glob import glob
+
+RDLogger.DisableLog("rdApp.*")
+
+def print_sys(s):
+    """system print
+
+    Args:
+        s (str): the string to print
+    """
+    print(s, flush=True, file=sys.stderr)
+
+
+def create_scaffold_split(df, seed, frac, entity):
+    """create scaffold split. it first generates molecular scaffold for each molecule and then split based on scaffolds
+    adapted from: https://github.com/mims-harvard/TDC/tdc/utils/split.py
+
+    Args:
+        df (pd.DataFrame): dataset dataframe
+        fold_seed (int): the random seed
+        frac (list): a list of train/valid/test fractions
+        entity (str): the column name for where molecule stores
+
+    Returns:
+        dict: a dictionary of splitted dataframes, where keys are train/valid/test and values correspond to each dataframe
+    """
+    random = Random(seed)
+
+    s = df[entity].values
+    scaffolds = defaultdict(set)
+    idx2mol = dict(zip(list(range(len(s))), s))
+
+    error_smiles = 0
+    for i, smiles in tqdm(enumerate(s), total=len(s)):
+        try:
+            scaffold = MurckoScaffold.MurckoScaffoldSmiles(
+                mol=Chem.MolFromSmiles(smiles), includeChirality=False
+            )
+            scaffolds[scaffold].add(i)
+        except:
+            print_sys(smiles + " returns RDKit error and is thus omitted...")
+            error_smiles += 1
+
+    train, val, test = [], [], []
+    train_size = int((len(df) - error_smiles) * frac[0])
```

Comment:
Why is `error_smiles` subtracted from all three splits?

Reply:
it is always subtracted from the length of the whole dataframe not from the individual splits length, to adjust the ratio of train-valid-test splits based on the ratio of error_smiles.

Reply:
yep, this seems ok

Reply:
ah, yeah, make sense, sorry, I missed the entire df.